### PR TITLE
[ci] Fix 2.1, 2.2 and 2.3 Dockerfiles

### PR DESCRIPTION
**Pull request by AlexJF**
**What does this PR do?**
Make Dockerfiles for 2.1, 2.2 and 2.3 work with debian bookworm (2.2 with bullseye due to https://github.com/ruby/setup-ruby/issues/496) rather than stretch, whose repos have been archived.

@ivoanjo pointed out we could have just rerouted the repos to debian archive but this way we stop relying on such old setups which makes things like docker setup and any eventual changes much easier since we're less likely to run into more unsupported/broken things.

**Motivation:**
Make github CI jobs work for 2.1, 2.2 and 2.3 again.

**Additional Notes:**

**How to test the change?**

**For Datadog employees:**
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: .circleci/images/primary/Dockerfile-2.1.10 [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # Note: See the "Publishing updates to images" note in ./README.md for how to publish new builds of this container image
 
 # taken from https://github.com/docker-library/ruby/blob/c5693b25aa865489fee130e572a3f11bccebd21b/2.1/Dockerfile
-FROM buildpack-deps:stretch AS ruby-2.1.10-stretch
+FROM buildpack-deps:bookworm AS ruby-2.1.10-bookworm
```
Comment (ivoanjo) [inline .circleci/images/primary/Dockerfile-2.1.10:7]: If anyone else is wondering [stretch is debian 9, and bookworm is debian 12](https://distrowatch.com/table.php?distribution=debian) so this buys us a lot of time! Nice :D

File: .circleci/images/primary/Dockerfile-2.2.10 [hunk 2]
```diff
@@ -4,8 +4,11 @@
 
 # Note: See the "Publishing updates to images" note in ./README.md for how to publish new builds of this container image
 
-# taken from https://github.com/docker-library/ruby/blob/b5ef401d348ca9b1d9f6a5cb4b25f32bf013daca/2.2/jessie/Dockerfile
-FROM buildpack-deps:stretch AS ruby-2.2.10-stretch
+# taken from https://hub.docker.com/layers/library/ruby/2.2.10/images/sha256-6c8e6f9667b21565284ef1eef81086adda867e936b74cc74a119ac380d9f00f9?context=explore
+# NOTE: Why bullseye and not bookworm like 2.1 and 2.3? Something fails with Ruby 2.2 compilation with newer images leading to segfaults and
+# errors like `Marshal.load reentered at marshal_load`: https://github.com/ruby/setup-ruby/issues/496#issuecomment-1520667671
+# These do not show up with bullseye and ubuntu-20.
+FROM buildpack-deps:bullseye AS ruby-2.2.10-bullseye
```
Comment (ivoanjo) [inline .circleci/images/primary/Dockerfile-2.2.10:11]: ...and bullseye is debian 11 ;)

File: .circleci/images/primary/Dockerfile-2.3.8 [hunk 3]
```diff
@@ -57,19 +131,6 @@ RUN DOCKERIZE_URL="https://github.com/powerman/dockerize/releases/download/v0.17
...
-# Install RubyGems and Bundler
-# NOTE: Rubygems 3.0.6 is the last version that seems to work fine in this image AND not drag bundler 2 along.
-# Later versions are either broken or they force the use of bundler 2, which we can't use because some of our
-# dependencies (e.g. rails 3.0.20) don't like it.
-RUN gem update --system '3.0.6'
-# Ruby 2.3 can support Bundler 2+
-# But hold back to < 2 for now, because some dependencies require it.
-RUN gem install bundler -v '1.17.3'
-ENV BUNDLE_SILENCE_ROOT_WARNING 1
-
```
Comment (ivoanjo) [inline .circleci/images/primary/Dockerfile-2.3.8:72]: I noticed that, unlike the other images, we're not installing a specific bundler version. This should probably be fine, but let's keep an eye out for it :)

**Review by ivoanjo — APPROVED** [review]
LGTM :+1: 

Thanks for picking this up!

Note that although images are built on every change, the new images are not yet published/used for CI.

TL;DR: Remember to go to https://github.com/DataDog/dd-trace-rb/actions/workflows/build-ruby.yml after merging this (or before?) and triggering a manual run with "Push images" selected, so the new images start getting used in CI.

**Comment by AlexJF** [comment]
> TL;DR: Remember to go to [DataDog/dd-trace-rb/actions/workflows/build-ruby.yml](https://github.com/DataDog/dd-trace-rb/actions/workflows/build-ruby.yml) after merging this (or before?) and triggering a manual run with "Push images" selected, so the new images start getting used in CI.

Yeah I'll follow the steps in https://github.com/DataDog/dd-trace-rb/blob/ab0601e51402a2db910f473bc43f780e751c4667/.circleci/images/primary/README.md now
